File: apps/api/internal/domain/locale/locale.go
// Package locale owns the Locale aggregate.
package locale

import (
	"context"
	"errors"

	"github.com/google/uuid"
	"golang.org/x/text/language"
)

// ErrInvalidCode is returned when a code is not a usable BCP 47 language tag.
var ErrInvalidCode = errors.New("locale: code must be a BCP 47 language tag (language[-script][-region][-variant]), e.g. 'de', 'de-CH', 'zh-Hant-TW' or 'es-419'")

// ErrInvalidLabel is returned when a label is empty or oversize.
var ErrInvalidLabel = errors.New("locale: label must be 1-50 characters")

// MaxCodeLength is the RFC 5646 §4.4.1 minimum buffer for a tag,
// and the width of every locale column.
const MaxCodeLength = 35

// Code is a canonical BCP 47 language tag identifying a locale.
//
// Canonical means RFC 5646 §4.5 canonicalization: deprecated subtags
// replaced by their preferred values (iw → he) and scripts the
// language implies anyway suppressed (en-Latn-US → en-US). Casing
// follows the RFC conventions (en-US, zh-Hant).
//
// Extensions (-u-, -t-) and private use (-x-) are rejected: they
// carry formatting preferences, not a distinct body of translations,
// so they don't belong in a locale's identity.
type Code string

// NewCode parses, validates and canonicalizes a tag.
func NewCode(s string) (Code, error) {
	tag, err := language.BCP47.Parse(s)
	if err != nil {
		return "", ErrInvalidCode
	}
	if base, _, _ := tag.Raw(); base.String() == "und" {
		return "", ErrInvalidCode
	}
	if len(tag.Extensions()) > 0 {
		return "", ErrInvalidCode
	}
	canonical := tag.String()
	if len(canonical) > MaxCodeLength {
		return "", ErrInvalidCode
	}
	return Code(canonical), nil
}

// String returns the underlying code value.
func (c Code) String() string { return string(c) }

// tag re-parses the code. Rows stored before canonicalization
// existed may hold non-canonical forms, so this goes through the
// same canonicalizer as NewCode.
func (c Code) tag() language.Tag {
	tag, _ := language.BCP47.Parse(string(c))
	return tag
}

// Language returns the primary language subtag, e.g. "zh" for zh-Hant-TW.
func (c Code) Language() string {
	base, _, _ := c.tag().Raw()
	return base.String()
}

// Script returns the explicit script subtag ("Hant" for zh-Hant-TW),
// or "" when the tag has none. Use [Code.Direction] for rendering
// decisions; it also accounts for the script a language implies.
func (c Code) Script() string {
	_, script, _ := c.tag().Raw()
	if script == (language.Script{}) {
		return ""
	}
	return script.String()
}

// Region returns the explicit region subtag ("TW", "419"), or "" when
// the tag has none.
func (c Code) Region() string {
	_, _, region := c.tag().Raw()
	if region == (language.Region{}) {
		return ""
	}
	return region.String()
}

// Direction is the base text direction of a locale.
type Direction string

// Direction values, matching the HTML dir attribute.
const (
	LTR Direction = "ltr"
	RTL Direction = "rtl"
)

// rtlScripts lists the ISO 15924 scripts whose characters are
// right-to-left (Unicode bidi class R or AL).
var rtlScripts = map[string]bool{
	"Adlm": true, "Arab": true, "Aran": true, "Armi": true, "Avst": true,
	"Chrs": true, "Cprt": true, "Elym": true, "Hatr": true, "Hebr": true,
	"Hung": true, "Khar": true, "Lydi": true, "Mand": true, "Mani": true,
	"Mend": true, "Merc": true, "Mero": true, "Narb": true, "Nbat": true,
	"Nkoo": true, "Orkh": true, "Ougr": true, "Palm": true, "Phli": true,
	"Phlp": true, "Phnx": true, "Prti": true, "Rohg": true, "Samr": true,
	"Sarb": true, "Sogd": true, "Sogo": true, "Syrc": true, "Thaa": true,
	"Yezi": true,
}

// Direction derives the text direction from the explicit script, or
// from the CLDR likely script when the tag has none (ar → Arab → rtl,
// pa → Guru → ltr, pa-Arab → rtl). Unknown scripts default to LTR.
func (c Code) Direction() Direction {
	script, confidence := c.tag().Script()
	if confidence != language.No && rtlScripts[script.String()] {
		return RTL
	}
	return LTR
}

// Matches reports whether raw names the same locale as c, comparing
// canonical forms, so "HE", "iw" and "he" all match Code("he").
func (c Code) Matches(raw string) bool {
	want, err := NewCode(raw)
	if err != nil {
		return false
	}
	if have, err := NewCode(string(c)); err == nil {
		return have == want
	}
	return string(c) == raw
}

// Label is a validated display label (e.g. "Deutsch", "English (US)").
type Label string

// NewLabel parses and validates a label literal.
func NewLabel(s string) (Label, error) {
	if l := len(s); l == 0 || l > 50 {
		return "", ErrInvalidLabel
	}
	return Label(s), nil
}

// String returns the underlying label value.
func (l Label) String() string { return string(l) }

// Locale is a single (project, language) combination.
type Locale struct {
	ID        uuid.UUID
	ProjectID uuid.UUID
	Code      Code
	Label     Label
	Enabled   bool
}

// Repository is the persistence port.
type Repository interface {
	Save(ctx context.Context, l Locale) error
	ListForProject(ctx context.Context, projectID uuid.UUID) ([]Locale, error)
	SetEnabled(ctx context.Context, id uuid.UUID, enabled bool) error
	Delete(ctx context.Context, id uuid.UUID) error
}

File: packages/sdk/src/locale.ts
// Locale identity helpers. Mirrors the Go side's locale.Code
// (apps/api/internal/domain/locale) so a locale means the same thing
// on every runtime. Locale data comes from the platform's Intl (CLDR);
// nothing here is a hand-maintained locale table except the list of
// right-to-left scripts, which is Unicode data that Intl only exposes
// through the not-yet-universal Intl.Locale#getTextInfo.

/** Base text direction, matching the HTML `dir` attribute. */
export type Direction = "ltr" | "rtl";

/** A locale's canonical tag, explicit subtags and text direction. */
export interface LocaleInfo {
  /** Canonical BCP 47 tag, e.g. `zh-Hant-TW`. */
  code: string;
  /** Primary language subtag, e.g. `zh`. */
  language: string;
  /** Explicit script subtag (`Hant`), `""` when the tag has none. */
  script: string;
  /** Explicit region subtag (`TW`, `419`), `""` when the tag has none. */
  region: string;
  /** Direction of the explicit or CLDR likely script. */
  direction: Direction;
}

// ISO 15924 scripts whose characters are right-to-left (Unicode bidi
// class R or AL). Keep in sync with rtlScripts in the Go locale package.
const RTL_SCRIPTS = new Set(
  (
    "Adlm Arab Aran Armi Avst Chrs Cprt Elym Hatr Hebr Hung Khar Lydi Mand Mani Mend " +
    "Merc Mero Narb Nbat Nkoo Orkh Ougr Palm Phli Phlp Phnx Prti Rohg Samr Sarb Sogd " +
    "Sogo Syrc Thaa Yezi"
  ).split(" "),
);

/**
 * Describe a BCP 47 tag. Accepts `_` as a separator (`en_US`) the way
 * the server does. Throws a `RangeError` for a malformed tag.
 */
export function describeLocale(tag: string): LocaleInfo {
  const loc = new Intl.Locale(tag.replaceAll("_", "-"));
  const likelyScript = loc.script ?? loc.maximize().script ?? "";
  return {
    code: loc.baseName,
    language: loc.language,
    script: loc.script ?? "",
    region: loc.region ?? "",
    direction: RTL_SCRIPTS.has(likelyScript) ? "rtl" : "ltr",
  };
}
